File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow = head
        fast = head
        last = None

        while fast.next and fast.next.next:
            slow = slow.next
            last = fast
            fast = fast.next.next


        curr = slow.next
        slow.next = None
        prev = None
        while curr != None:
            nextNode = curr.next
            curr.next = prev
            prev = curr
            curr = nextNode
        #print(prev.val,prev.next.val)

        pointer = head
        fin = og = ListNode()
        while prev and pointer:
            fin.next = pointer
            pointer = pointer.next
            fin = fin.next
            fin.next = prev
            prev = prev.next
            fin = fin.next

        if prev == None:
            fin.next = pointer
        else:
            fin.next = prev

        #return og.next

        

        

        
        

        

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* og = new ListNode();
        og->next = head;
        ListNode* prevGroup = og;

        while(true){
            ListNode* kth =prevGroup;
            for (int i = 0; i < k && kth != nullptr; i +=1){
                kth = kth ->next;
            }
            if (kth == nullptr){break;}

            ListNode* groupNext = kth->next;
            ListNode* prev = groupNext;
            ListNode* curr = prevGroup->next;
            while (curr != groupNext) {
                ListNode* nxt = curr->next;
                curr->next = prev;
                prev = curr;
                curr = nxt;
            }
            

            ListNode* newPrevGroup = prevGroup->next;
            prevGroup->next = kth;
            prevGroup = newPrevGroup;
        }
        return og->next;
    }
};
